File: src/network_identity/fingerprint_policy.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum FingerprintFieldPriorityLayer {
    L1,
    L2,
    L3,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum FingerprintPerfBudgetTag {
    Light,
    Medium,
    Heavy,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct FingerprintFieldPriorityRule {
    pub field: &'static str,
    pub layer: FingerprintFieldPriorityLayer,
    pub reason: &'static str,
}

pub fn fingerprint_field_priority_rules() -> Vec<FingerprintFieldPriorityRule> {
    vec![
        FingerprintFieldPriorityRule { field: "user_agent", layer: FingerprintFieldPriorityLayer::L1, reason: "must be truly consumed first" },
        FingerprintFieldPriorityRule { field: "accept_language", layer: FingerprintFieldPriorityLayer::L1, reason: "must be truly consumed first" },
        FingerprintFieldPriorityRule { field: "locale", layer: FingerprintFieldPriorityLayer::L1, reason: "must be truly consumed first" },
        FingerprintFieldPriorityRule { field: "timezone", layer: FingerprintFieldPriorityLayer::L1, reason: "must be truly consumed first" },
        FingerprintFieldPriorityRule { field: "viewport", layer: FingerprintFieldPriorityLayer::L1, reason: "must be truly consumed first" },
        FingerprintFieldPriorityRule { field: "platform", layer: FingerprintFieldPriorityLayer::L1, reason: "must be truly consumed first" },
        FingerprintFieldPriorityRule { field: "client_hints", layer: FingerprintFieldPriorityLayer::L2, reason: "high-value next layer" },
        FingerprintFieldPriorityRule { field: "hardware_concurrency", layer: FingerprintFieldPriorityLayer::L2, reason: "high-value next layer" },
        FingerprintFieldPriorityRule { field: "device_memory", layer: FingerprintFieldPriorityLayer::L2, reason: "high-value next layer" },
        FingerprintFieldPriorityRule { field: "color_scheme", layer: FingerprintFieldPriorityLayer::L2, reason: "high-value next layer" },
        FingerprintFieldPriorityRule { field: "canvas", layer: FingerprintFieldPriorityLayer::L3, reason: "advanced fingerprint layer" },
        FingerprintFieldPriorityRule { field: "webgl", layer: FingerprintFieldPriorityLayer::L3, reason: "advanced fingerprint layer" },
        FingerprintFieldPriorityRule { field: "audio", layer: FingerprintFieldPriorityLayer::L3, reason: "advanced fingerprint layer" },
        FingerprintFieldPriorityRule { field: "fonts", layer: FingerprintFieldPriorityLayer::L3, reason: "advanced fingerprint layer" },
        FingerprintFieldPriorityRule { field: "anti_detection_flags", layer: FingerprintFieldPriorityLayer::L3, reason: "advanced fingerprint layer" },
    ]
}

pub fn classify_fingerprint_field(field: &str) -> Option<FingerprintFieldPriorityLayer> {
    fingerprint_field_priority_rules()
        .into_iter()
        .find(|rule| rule.field == field)
        .map(|rule| rule.layer)
}

pub fn default_fingerprint_perf_budget_for_layer(layer: FingerprintFieldPriorityLayer) -> FingerprintPerfBudgetTag {
    match layer {
        FingerprintFieldPriorityLayer::L1 => FingerprintPerfBudgetTag::Light,
        FingerprintFieldPriorityLayer::L2 => FingerprintPerfBudgetTag::Medium,
        FingerprintFieldPriorityLayer::L3 => FingerprintPerfBudgetTag::Heavy,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn priority_rules_cover_expected_layers() {
        let rules = fingerprint_field_priority_rules();
        assert!(rules.iter().any(|r| r.field == "user_agent" && r.layer == FingerprintFieldPriorityLayer::L1));
        assert!(rules.iter().any(|r| r.field == "hardware_concurrency" && r.layer == FingerprintFieldPriorityLayer::L2));
        assert!(rules.iter().any(|r| r.field == "canvas" && r.layer == FingerprintFieldPriorityLayer::L3));
    }

    #[test]
    fn classify_field_returns_expected_layer() {
        assert_eq!(classify_fingerprint_field("timezone"), Some(FingerprintFieldPriorityLayer::L1));
        assert_eq!(classify_fingerprint_field("client_hints"), Some(FingerprintFieldPriorityLayer::L2));
        assert_eq!(classify_fingerprint_field("audio"), Some(FingerprintFieldPriorityLayer::L3));
        assert_eq!(classify_fingerprint_field("unknown_field"), None);
    }

    #[test]
    fn default_perf_budget_matches_layer_weight() {
        assert_eq!(default_fingerprint_perf_budget_for_layer(FingerprintFieldPriorityLayer::L1), FingerprintPerfBudgetTag::Light);
        assert_eq!(default_fingerprint_perf_budget_for_layer(FingerprintFieldPriorityLayer::L2), FingerprintPerfBudgetTag::Medium);
        assert_eq!(default_fingerprint_perf_budget_for_layer(FingerprintFieldPriorityLayer::L3), FingerprintPerfBudgetTag::Heavy);
    }
}

File: backend/internal/browser/fingerprint_policy.go
package browser

// FingerprintFieldPriorityLayer classifies fingerprint fields by importance.
type FingerprintFieldPriorityLayer int

const (
	FPLayerL1 FingerprintFieldPriorityLayer = iota // must be consumed first
	FPLayerL2                                       // high-value next layer
	FPLayerL3                                       // advanced fingerprint layer
)

// FingerprintPerfBudgetTag maps layers to performance budgets.
type FingerprintPerfBudgetTag int

const (
	FPBudgetLight  FingerprintPerfBudgetTag = 0
	FPBudgetMedium FingerprintPerfBudgetTag = 1
	FPBudgetHeavy  FingerprintPerfBudgetTag = 2
)

// FingerprintFieldPriorityRule maps a field to its priority layer with a reason.
type FingerprintFieldPriorityRule struct {
	Field  string
	Layer  FingerprintFieldPriorityLayer
	Reason string
}

// FingerprintFieldPriorityRules returns the canonical priority rules for all 15 fingerprint fields.
func FingerprintFieldPriorityRules() []FingerprintFieldPriorityRule {
	return []FingerprintFieldPriorityRule{
		{Field: "user_agent", Layer: FPLayerL1, Reason: "must be truly consumed first"},
		{Field: "accept_language", Layer: FPLayerL1, Reason: "must be truly consumed first"},
		{Field: "locale", Layer: FPLayerL1, Reason: "must be truly consumed first"},
		{Field: "timezone", Layer: FPLayerL1, Reason: "must be truly consumed first"},
		{Field: "viewport", Layer: FPLayerL1, Reason: "must be truly consumed first"},
		{Field: "platform", Layer: FPLayerL1, Reason: "must be truly consumed first"},
		{Field: "client_hints", Layer: FPLayerL2, Reason: "high-value next layer"},
		{Field: "hardware_concurrency", Layer: FPLayerL2, Reason: "high-value next layer"},
		{Field: "device_memory", Layer: FPLayerL2, Reason: "high-value next layer"},
		{Field: "color_scheme", Layer: FPLayerL2, Reason: "high-value next layer"},
		{Field: "canvas", Layer: FPLayerL3, Reason: "advanced fingerprint layer"},
		{Field: "webgl", Layer: FPLayerL3, Reason: "advanced fingerprint layer"},
		{Field: "audio", Layer: FPLayerL3, Reason: "advanced fingerprint layer"},
		{Field: "fonts", Layer: FPLayerL3, Reason: "advanced fingerprint layer"},
		{Field: "anti_detection_flags", Layer: FPLayerL3, Reason: "advanced fingerprint layer"},
	}
}

// ClassifyFingerprintField returns the priority layer for a fingerprint field, or nil if unknown.
func ClassifyFingerprintField(field string) *FingerprintFieldPriorityLayer {
	for _, rule := range FingerprintFieldPriorityRules() {
		if rule.Field == field {
			layer := rule.Layer
			return &layer
		}
	}
	return nil
}

// DefaultPerfBudgetForLayer maps a priority layer to its default performance budget tag.
func DefaultPerfBudgetForLayer(layer FingerprintFieldPriorityLayer) FingerprintPerfBudgetTag {
	switch layer {
	case FPLayerL1:
		return FPBudgetLight
	case FPLayerL2:
		return FPBudgetMedium
	case FPLayerL3:
		return FPBudgetHeavy
	default:
		return FPBudgetLight
	}
}

// LayerForPerfBudget returns the recommended field layer that fits within the given budget.
// Light budget → L1 only, Medium → L1+L2, Heavy → all fields.
func LayerForPerfBudget(budget FingerprintPerfBudgetTag) FingerprintFieldPriorityLayer {
	switch budget {
	case FPBudgetLight:
		return FPLayerL1
	case FPBudgetMedium:
		return FPLayerL2
	default:
		return FPLayerL3
	}
}
